src/AI/JakeMan.java

```diff
@@ -846,29 +1107,122 @@ else if( um == copter ) // Consider downgrading a tank to an inf to get a copter
     return builds;
   }
 
-  /**
-   * Sort units by funds amount in descending order.
-   */
-  private static class UnitModelFundsComparator implements Comparator<Entry<UnitModel, Double>>
-  {
-    @Override
-    public int compare(Entry<UnitModel, Double> entry1, Entry<UnitModel, Double> entry2)
-    {
-      double diff = entry2.getValue() - entry1.getValue();
-      return (int) (diff * 10); // Multiply by 10 since we return an int, but don't want to lose the decimal-level discrimination.
-    }
-  }
-
   private void init(GameMap map)
   {
     infantry = myArmy.cos[0].getUnitModel(UnitModel.TROOP);
     allTanks = myArmy.cos[0].getAllModels(UnitModel.ASSAULT);
+    tank     = allTanks.get(0);
+    mdTank   = allTanks.get(1);
     antiAir  = myArmy.cos[0].getUnitModel(UnitModel.SURFACE_TO_AIR);
     copter   = myArmy.cos[0].getUnitModel(UnitModel.ASSAULT | UnitModel.AIR_LOW, false);
+
+    counterBuildSetup();
+
+    allCounterContestants = new HashSet<>();
+    for( UnitModel counterable : unitTypeToCounterMap.keySet() )
+    {
+      allCounterContestants.add(counterable);
+      for( CounterRatio ratio : unitTypeToCounterMap.get(counterable) )
+        allCounterContestants.add(ratio.counter);
+    }
```

Comment:
You are adding both the type we want to counter, and the type that counters it? And then the only place this is used is to ignore cumulative health for units that aren't in this map? So it's more like a cache of units that Jake is willing to think about?
Also it seems like this might have overlap with `unitTypeToCounterMap`, which is also used in `queueUnitProductionActions`

Reply:
> You are adding both the type we want to counter, and the type that counters it?

Yes. This is a cache for the types of units we count for counterbuild logic purposes (vs the ones we ignore, like tanks and infantry).
> And then the only place this is used is to ignore cumulative health for units that aren't in this map?

No, see previous line.
> So it's more like a cache of units that Jake is willing to think about?

Yes, actually. It's the set of units that factor into his special/extra planning.
> Also it seems like this might have overlap with `unitTypeToCounterMap`, which is also used in `queueUnitProductionActions`

It intentionally overlaps, but I have to count both things that I might *want* to counter and might *build as* a counter.
Without both categories, JakeMan will either not counterbuild at all, or he'll just keep spamming AA as long as you have a jet or two. ...I say this from experience.